BUG: Revert multifield-indexing adds padding bytes for NumPy 1.15.

ahaldane opened the pull request:
Fixes #10387
Fixes #10344 

This is a potential fix to the multifield-indexing bugs in 1.14.0 caused by how easy it is now to get a structured array with padding bytes, which trigger other old bugs like #3176, #8100.

Changed Behavior
-------------------------------

This PR reverts the change in 1.14.0 that a multifield-index returns a view instead of a copy, but we plan to make the change in 1.15. That is:

```python
>>> a = np.zeros(3, dtype=[('a', 'i4'), ('b', 'i4'), ('c', 'f4')])
```
1.13 behavior:
```python  
>>> a[['a', 'c']] 
array([(0, 0.), (0, 0.), (0, 0.)],
      dtype=[('a', '<i4'), ('c', '<f4')]
```
1.14.0 behavior (pushed off to 1.15):
```python
>>> a[['a', 'c']]   
array([(0, 0.), (0, 0.), (0, 0.)],
      dtype={'names':['a','c'], 'formats':['<i4','<f4'], 'offsets':[0,8], 'itemsize':12})
```

Summary of the problem
--------------------------------

Why revert? The problem boils down to padding bytes. Compare the 1.13 to 1.14.0 output above, and note how there are "padding" or missing bytes at offset 4 in the structure in 1.14.0, and how the itemsize is different. The 1.13 behavior essentially returns a copy of the 1.14 output, but with the fields re-packed so that they are contiguous. The two ways users are affected were: 

 A) Code that attempts to "view" the result, as in:

         >>> a[['a', 'c']].view('i8')
         array([0, 0, 0])

In 1.13 this works fine, since the indexed array has an      itemsize of 8, but in 1.14.0 this now fails since the       itemsize of the view is still 12.

  B) Code that attempts to use np.load/np.save to save these       arrays, as in:

         >>> np.save('file', a[['a', 'c']])
         >>> np.load('file')
       
This fails because of another long-standing bug, that       currently np.load cannot handle structured arrays with       padding bytes, https://github.com/numpy/numpy/issues/8100.  In 1.13 this bug was not triggered because there were no       padding bytes, but in 1.14.0 they now appear often.

Proposed Way Forward:
-------------------------------

Revert the copy-> view change, and leave it for 1.15. 

That way, we have time to fix #8100 by then, solving problem B). 

Problem A) will still cause broken code, but we hope to mitigate this by introducing a new method `pack_fields` in this PR for 1.14.1, which allows convertsion of the 1.15 output into the 1.13 output:
```python
>>> np.pack_fields(a[['a', 'c']])
array([(0, 0.), (0, 0.), (0, 0.)],
      dtype=[('a', '<i4'), ('c', '<f4')]
```

See comments below for more discussion. Also see the "documentation" updates at the end of the diff for a description of the proposed behavior, under "Accessing Multiple Fields".

---

<details>
<summary>(Old outdated summary, for reference)</summary>

This PR implements a couple different possible behaviors and makes them user-configurable; we can decide in discussion which one we want, which should be default, or which to remove. You control the behavior by setting the variable `np.multifield_index_setting` to one of `'view'`, `'copy'` or `'copywarn'` (default).

For `'view'`, the behavior is unchanged from 1.14.0, meaning that multi-field indexing returns a view which will often contain padding bytes.

For `'copy'`, the behavior reverts back to 1.13 behavior, in which a packed-copy of the array fields is returned. This should solve most of people's bugs.

For `'copywarn'`, do the same as `'copy'` but warn anytime multifield indexing is used. (default)

Is this kind of configuration option a good idea, or will it lead to a dependency mess? (I got the idea from the legacy mode for printing, which also allows the old and new behavior but defaults to the new one, unlike this PR currently).

In this PR I also introduce a new user-visible function `np.pack_fields`, which removes padding bytes from structured ndarrays and dtypes. Its existence should help any users transition to the new behavior: Doing `np.pack_fields(a[['a', 'c']])` in the new behavior (`'view'` setting) returns what you get in the old (`'copy'`) setting.

</details>

eric-wieser commented on numpy/core/numeric.py:
```diff
@@ -623,6 +624,45 @@ def asfortranarray(a, dtype=None):
+    packed : ndarray or dtype
+       Copy of a with padding bytes removed
+    """
+    if isinstance(a, dtype):
```
How about `if not isinstance(a, dtype): return a.astype(pack_fields(a.dtype))`

eric-wieser commented on numpy/core/numeric.py:
```diff
@@ -623,6 +624,45 @@ def asfortranarray(a, dtype=None):
     return array(a, dtype, copy=False, order='F', ndmin=1)
 
+def pack_fields(a):
+    """Return copy of a structured array with padding between fields removed.
```
And fields reordered in memory

ahaldane replied:
Oh right I forgot about reordering. Will think about it tomorrow.

eric-wieser commented:
I'd rather not introduced any more global state into numpy, to avoid running into #9444 in more places.

ahaldane commented:
Here's a few different non-configurable ways we could go forward, which I would be OK with:

1. Keep the new `view` behavior in 1.14, and tell people to use `pack_fields` if their code broke. (statsmodels people may be unhappy... but I ran their tests on their master, I see only 1 test failure before this PR).
2.  Revert to `copywarn` behavior for 1.14 with a note to use `pack_fields` (it will be a noop for copywarn), and then switch to solution 1 in 1.15. Needs a way to turn off the warning.
3. Keep the `copy` behavior forever for back-compatibility. Honestly I'm not even that unhappy with that. I care more about all the other #6053 changes which we are keeping.

eric-wieser commented:
> Keep the copy behavior forever for back-compatibility. Honestly I'm not even that unhappy with that.

That actually doesn't seem too awful. It would be even better if we could preserve the view behavior under a different API, perhaps:
```
a[['field1', 'field2']]  # master creates a view, proposal copies
a.view(a.dtype[['field1', 'field2']])  # a view no matter what
```

ahaldane commented:
I want to collect some past discussions on this issue:

We discussed multifield-views more extensively in #350 and briefly in #5994 and #3641. This change has been originally planned as far back as numpy 1.7. It was mentioned as a `FutureWarning` in the 1.8 release note.

I also found these mailing list threads. ★ means more important.

 * ★ [ (Feb 05, 2009) Selection of only a certain number of fields ](http://numpy-discussion.10968.n7.nabble.com/Selection-of-only-a-certain-number-of-fields-td18047.html) Longer discussion started by Travis Oliphant introducing multifield indexing, and debating view vs copy. People, including Travis, prefer "view", but travis later argues there may be a technical challenge which forces us to "copy". (This is overcome today as of #6053).
 *  [FYI: New "select-multiple-fields" behavior (Feb 11, 2009)](http://numpy-discussion.10968.n7.nabble.com/FYI-New-quot-select-multiple-fields-quot-behavior-td18862.html) Travis thinks is is not possible to make multifield indexing return a view. He also notes that fancy indexing does not return a view either.
 * [issue #1431 accessing multiple fields in a recarray (Jun 27, 2010) ](http://numpy-discussion.10968.n7.nabble.com/issue-1431-accessing-multiple-fields-in-a-recarray-td14998.html) Somewhat unrelated, this is a very short discussion about field reordering in multifield indexing. #6053 supersedes/fixes this.
 * [view of a structured array? ( Jan 25, 2012) ](http://numpy-discussion.10968.n7.nabble.com/view-of-a-structured-array-td29630.html) Question asking if it is possible to get a view from a multi-field index. Answer is no, workaround suggested.
 *  ★ [Extracting sub-fields from an array as a view (PR 350) ( Jul 14, 2012)](http://numpy-discussion.10968.n7.nabble.com/Extracting-sub-fields-from-an-array-as-a-view-PR-350-td31202.html) Travis proposes to turn the copy into a view. Presumably this discussion turned into #350.
 * [About ready to start 1.8 release process. (Aug 12, 2013)](http://numpy-discussion.10968.n7.nabble.com/About-ready-to-start-1-8-release-process-td34776.html) Chuck asked whether we wanted to go ahead with multifield-views FutureWarning, but no-one answered. So we did.

I think if that we want to argue in favor of the change, there is clearly a long history of planning and warning about it.

eric-wieser commented:
While there is a history of planning and warning, it seems that support for the non-contiguous dtypes it creates was a gap in our preparation - depending on how much more we want to squeeze into 1.14.1, it might be better to roll back views for now to give us time to think about the right way to fix this problem in 1.15.

ahaldane commented:
Yes, that is what I'm thinking too. 

The problems caused by the new behavior really boil down to two issues: #10409 for views, and #10387 for save/load.

If we delay this change until later, I am pretty sure we can fix #10387 so that it will not be a problem. I'd guess #10387 is the one causing the most test failures, too.

But I think #10409 will cause unavoidable downstream code-breakage if we return a padded view. The best we can do is introduce `pack_fields` now, and perhaps add more warnings so people start using it. I suspect that this problem is the rarer one.

ahaldane commented:
Updated. Now we keep the copy behavior for 1.14, to be changed for real in 1.15, hopefully. 

I added back all the deprecation warning tests. By the way, those people who ran across the view issue of #10409 have quite surely been getting big FutureWarnings from numpy. No excuses!

In these latest changes I kept a simplified version of the `np.multifield_index_view` setting, but made it mostly "secret". See the new documentation section in this PR for details.

Test failure is not real, it's due to #10425

eric-wieser commented on doc/release/1.14.1-notes.rst:
```diff
@@ -0,0 +1,22 @@
+==========================
+NumPy 1.14.1 Release Notes
```
Extracted to #10431

ahaldane commented:
If there are no objections from devs at this point, later today or tomorrow I'll post to the list about this PR and how we are reverting and pushing the copy->view change to 1.15.

I don't expect anyone will object, but maybe users who were affected have extra ideas we could implement.

rgommers commented:
Disclaimer: I've followed along but didn't have enough time to read all the history in linked issues/threads.

Opinions:
- Agree with @eric-wieser that more global state is not desirable
- The revert seems necessary
- In general, trying to change copy behavior to view behavior is a bad idea, the gains are not worth the backwards compatibility breaks. (for new functionality, by all means go for views of course)
- In this case, there's something fishy in addition to that - if a new function `pack_fields` is needed to guard against mishaps with the view behavior, that seems like a sign that keeping the copy behavior is the best option from an API perspective.

So if you still want to make the change in 1.15.0, a clear story about the pros/cons of just that change would be good to see.

Also, was introducing `np.pack_fields` proposed on the mailing list? It may be a useful utility, but doesn't seem to deserve being in the top-level namespace.

ahaldane commented:
Yes, we got rid of the global state (except... I kept it in hidden form for dev purposes), and are planning to revert. We can also move `np.pack_fields` to the `np.rec` submodule which has similar methods.

I'll incorporate your other points into the email this weekend.

charris commented:
IIRC, I did do a preliminary run at replacing the deprecation with a view several releases ago, but ran into the same problem of filler bytes and a failing test, so never followed through. Another thing we should revisit is the diagonal view replacing a copy.

eric-wieser commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -1557,11 +1570,50 @@ _get_field_view(PyArrayObject *arr, PyObject *ind, PyArrayObject **view)
+            return 0;
+        }
+
+        /*
```
I think everything from here down should go in a helper function with a name like (but better than) `maybe_copy_view(view)`

eric-wieser commented on numpy/core/numeric.py:
```diff
@@ -623,6 +624,66 @@ def asfortranarray(a, dtype=None):
+    byte offsets.
+
+    This method also accepts a dtype object, in which case it returns
+    a dtype with repacked fields.
```
I wonder if we want to name this `dtype.packed_fields()`, or some method of dtype objects

ahaldane replied:
Yeah that would avoid polluting the `np` namespace.  

But I do have an aesthetic objection: I don't like that OO style, of making it the attribute of a class: Being used to C, I think toplevel functions are more pleasing than (class) methods, and my impression is numpy is usually organized that way too. Most other dtype related functions are functions, not dtype methods: https://docs.scipy.org/doc/numpy-1.13.0/reference/routines.dtype.html. I always use `np.sum(a)`, not `a.sum()`. 

`pack_fields` also happens to be a "pure function" in the functional programming sense, somehow justifying to me it should not be a class method. I can support putting it in a namespace (not a class) like `np.lib.recfunctions`, `np.rec` or somewhere else though.

PS. I am going to rename it `repack_fields(a, align=False)`.

eric-wieser replied:
> Most other dtype related functions

ahaldane commented:
Updated following all comments.

`repack_fields` now lives with its kindred in `numpy.lib.recfunctions`.

eric-wieser commented on numpy/lib/recfunctions.py:
```diff
@@ -807,6 +807,209 @@ def repack_fields(a, align=False, recurse=True):
+    fields = []
+    for name in dt.names:
+        sdt = dt.fields[name]
+        count = 1 if sdt[0].shape == () else np.product(sdt[0].shape)
```
Why the `if`?

ahaldane replied:
Oh yeah I guess product works on empty tuples.

eric-wieser commented:
Scope creep happening here. Can we split off the "lots of functions for 1.15" and "revert the 1.14 behavior" parts of this?

ahaldane commented:
All right, removed the last commit, will resubmit it as a separate PR.

eric-wieser commented on numpy/lib/recfunctions.py:
```diff
@@ -731,6 +731,81 @@ def rec_append_fields(base, names, data, dtypes=None):
     return append_fields(base, names, data=data, dtypes=dtypes,
                          asrecarray=True, usemask=False)
 
+def repack_fields(a, align=False, recurse=True):
```
I think the default should maybe be `recurse=False`

eric-wieser commented on numpy/lib/recfunctions.py:
```diff
@@ -731,6 +731,81 @@ def rec_append_fields(base, names, data, dtypes=None):
+
+    return np.dtype({'names': a.names,
+                     'formats': formats,
+                     'titles': titles}, align=align)
```
This changes the data type of union structured arrays, such as `(int, [('a', int)])`, but more importantly `(np.record, [...])`.

eric-wieser replied:
Fixed with `np.dtype((a.type, {...}))` (plus a comment and test for `np.record`), I think

eric-wieser commented on numpy/lib/recfunctions.py:
```diff
@@ -731,6 +731,81 @@ def rec_append_fields(base, names, data, dtypes=None):
+                   if f[0].names else f[0] for f in fields]
+    else:
+        formats = [f[0] for f in fields]
+    titles = [f[2] if len(f) == 3 else None for f in fields]
```
I think this might be clearer as a manual for loop that appends all three in parallel

eric-wieser commented on numpy/lib/recfunctions.py:
```diff
@@ -731,6 +731,81 @@ def rec_append_fields(base, names, data, dtypes=None):
+    dtype([('f0', 'u1'), ('f1', '<i4'), ('f2', '<f8')])
+    >>> print_offsets(packed_dt)
+    ('offsets:', [0, 1, 5])
+    ('itemsize:', 13)
```
I'd be inclined to show the py3 output here, which is probably what you intended anyway

eric-wieser commented on numpy/lib/recfunctions.py:
```diff
@@ -731,6 +731,81 @@ def rec_append_fields(base, names, data, dtypes=None):
+
+    """
+    if not isinstance(a, np.dtype):
+        return a.astype(repack_fields(a.dtype, align=align, recurse=recurse))
```
Should this have `copy=False`?

eric-wieser reviewed (commented):
I think I'd still prefer the operation to be called `dtype.to_packed()` or similar - possibly still providing the recfunctions function as a helper for array arguments

eric-wieser commented:
In fact, rather than rushing out `repack_fields`, perhaps we should leave it hidden in 1.14.1, and have a mailing list discussion about naming things for 1.15?

ahaldane commented:
I was sort of hoping to make it public in 1.14.1 so that downstream projects would have a release cycle to update code. 

If we only make these funcs public in 1.15, I think we should then push off the copy->view change to 1.16.

I guess the question is, do we need discussion on the mailing list, or is it pretty clear what it should do? Admittedly there was a lot of debate on the list so far. On the other hand, I think the desired behavior is fairly clear.

eric-wieser commented:
> downstream projects would have a release cycle to update code using it.

I suppose this also justifies your global flag to turn this on and off - they can use that in their tests

> Admittedly there was a lot of debate on the list so far. On the other hand, I think the desired behavior is fairly clear.

There seemed to be a lot of confusion about recarray/structured-arrays/what-actually-changed on the mailing list, and I'm not sure much of it was useful. I agree that the semantics of the function you propose are fairly clear. My main concerns are:

* Putting key functionality in `recfunctions`:
  * Do we really want to further the struct/record array confusion by adding more to this?
  * `recfunctions` is still not in the public docs
* Having it be overloaded by dtype/array - there's not a huge amount of precedent for this, and it makes the work in `numpy_stubs` a little harder. But clearly we need to expose a dtype converter somewhere so that users can do other things with a packed type like allocate an empty array to store results in.

ahaldane commented:
True, I'm not decided what to do about recfunctions, whether we should move it or more officially document it. For back-compat, I think it will have to be there forever, but we might move the methods somewhere else and just link them from numpy.lib.recfunctions. In that sense, there is no harm in putting `repack_fields` there today, since we can always move it later and just make it accessible from both places.

By overloading, you mean making it a class method `dtype.repack()`? I still think field-packing a specialized-enough case that we can just leave it as a semi-hidden method in `numpy.lib.recfunctions`. It will be confusing to people who have never heard of structured arrays to have it appear as a method on all dtypes.  (And at least after this PR I refer to it in docs.)

charris commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -1383,15 +1383,75 @@ array_subscript_asarray(PyArrayObject *self, PyObject *op)
+     * help development in the 1.14 to 1.15 transition.
+     * If true, return a view instead of a copy.
+     */
+    mf_setting = npy_import("numpy", "multifield_index_view");
```
How do you plan to use this? Are you going to modify `multifield_index_view` on the fly? I'm skeptical of this approach.

ahaldane replied:
Yes, it can be modified on the fly, but it is only for testing purposes. See the rewritten section of the docs, and also the unit tests that use it.

charris replied:
Still, it introduces unfortunate global state. It would be preferable to have a simple PR that fixes the problem for 1.14.1 and then deal with 1.15.

ahaldane replied:
No problem, I'll remove it.

charris commented:
This PR is against master. I assume this restores previous behavior there also with intent to redo things later?

ahaldane commented:
Don't review this quite yet, I have changes suggested by Eric which I haven't pushed; will do tonight.

ahaldane commented:
And yes, it restores the previous behavior with the intent to redo later. (Where "redo" means to delete a block of code). It also updates the docs to better warn about the changes.

ahaldane commented:
All right. What I've done is removed all of the code that should only be in the 1.14.1 branch from this PR, and opened a new PR for 1.14.1 in #10537. So this PR now contains the "final" state I am aming for in 1.15.

If we make changes to `repack_fields` in either PR, I will port the changes to the other PR before we merge.

charris commented:
@ahaldane A rebase would probably fix the circleci failure.  I assume we still want this for 1.15.

ahaldane commented:
Before we can finish this, we need to fix the problems in np.load in #8100. I was planning on more time to address that one before 1.15, but I'll take a look at that next to try and get it done in time.

mhvk commented:
I'm very late to this discussion and only noticed it again because it was recently changed, but anyway: I still like views very much, but the `repack_fields` function feels a bit clunky. At some level, the problem it intends to solve is very similar to taking a non-contiguous slice, and "repacking" that, and for slices that is what making a simple copy would do. Would that be possible? (I fear it might well break even more code if we started doing that for structured arrays, but thought I'd ask anyway...)

A propos copying, though, do you have a specific solution in mind for the `np.save` problem? In particular, is the plan to store the padding as well? I ask in part since for slicing `np.save` just writes the bytes that actually matter, and presumably the same should happen for structured dtypes, which means you'd need some type of packing already.

mhvk commented:
@ahaldane - I guess my comment was in part a question of whether those padding bytes should in fact be stored...

ahaldane commented:
That's worth considering... my initial reaction is that we should keep the padding bytes, since a save/load pair shouldn't modify the memory layout of an array, since people might depend on a particular memory layout.

mhvk commented:
@ahaldane - it would be the memory layout of the dtype that would be the only risk; the array layout of slices is changed. I guess my naive expectation would certainly be that padding bytes do not get stored; the main question would be the extent to which they get reinserted.

ahaldane commented:
Just to motivate my thought more, consider that one of the main purposes of structured arrays (at least, the purpose I see and have written up in the structured array docs) is to mimic C structs for the purpose of interfacing with C programs and data. I have been discouraging use of structured arrays for tabular data analysis where the "fields" act more like simply an extra dimension.

Fot that C struct-like purpose, for instance, we have an "aligned" memory layout mode which adds padding bytes to mimic the C struct. This way `.tobytes()` gives the same padding as C, and the itemsize is the same as in C. So someone could save a structured numpy array, then read it back in with `np.load`, and then pass it to a C library (through a python C extension). If we drop the padding that would not be possible.

On the other hand, I agree that if we think of a structured array as a table it makes some sense to remove the padding bytes. It's also possible that I have been pushing too hard against that use and in favor of C struct behavior.

charris commented:
C structures are notorious for being non-portable on account of the alignment of the elements depending on platform and compiler. I always tried to avoind them for that reason and when I did use them, was careful to lay them out so that alignment would not an issue. 

In the current situation, I think we should choose a preferred method if possible. If we need to interact with C code, It might be best to store compressed and load with an alignment suitable to the platform, although I'm not sure how we determine that. 

On the question of view vs copies, I rather like views, but they do cause backwards compatibility problems, so in this particular case it might be worth having a setable option. 

Similar to this problem, we also had the proposal to make `ndarray.diagonal` return views, which it now does, but we haven't made the view writable and it has been hanging on unresolved since 1.9. It would be good to resolve that issue.

I don't use recarrays, so I would prefer folks who do to make the decision how to proceed.

mhvk commented:
@ahaldane - my comment was mostly about *saving* the array compressed (as Chuck notes); whether one re-expands after is at some level a separate question. I use the structured dtypes myself mostly to keep elements together (in simplest form, xyz of vectors without having a shape ending in 3).

ahaldane commented:
As for how I am thinking of implementing all this, note that what `numpy/lib/npyio.py` and `numpy/lib/format.py` do is hijack the PEP3118 format description, and serialize the array using it.

I think the problem in #8100 isn't with serializing the *data*, but with serializing the *dtype*.

The fix to #8100 will be either 1. to rewrite format.py so it uses its own code instead of relying on the pep3118 `dtype.descr` to save the dtype, or 2. Always interpreting empty field names `''` in a pep3118 descr as a padding region which should not be given a name. See these comments: https://github.com/numpy/numpy/issues/3176#issuecomment-125773238
https://github.com/numpy/numpy/issues/7797#issuecomment-230186999

One way or another, we need a way to reliably save/store the dtype object itself, in a way that we can restore the padding byte info. We don't have that now.

There is a comment in `numpy/lib/format.py` about structured types and save/load:

```
.. warning::

  Due to limitations in the interpretation of structured dtypes, dtypes
  with fields with empty names will have the names replaced by 'f0', 'f1',
  etc. Such arrays will not round-trip through the format entirely
  accurately. The data is intact; only the field names will differ. We are
  working on a fix for this. This fix will not require a change in the
  file format. The arrays with such structures can still be saved and
  restored, and the correct dtype may be restored by using the
  ``loadedarray.view(correct_dtype)`` method.
```

and note the definition of `dtype_to_descr` there, which is a hack to reprocess descrs so they can be accepted by the `dtype` function.

The additional detail that we might not want to actually write the padding data to file seems like a simple extra change, once we have the right behavior for saving/parsing the dtype. We would just use `repack_fields` to get the packed version of the dtype, and save/load using it, finally converting back to the unpacked form before returning to the user.

Incidentally, looking at all this again it's not clear to me why we don't just serialize the dtype using the pep3118 format string itself, as given by `memoryview(arr).format`. Maybe it's because the pep3118 interface is still somewhat buggy (it still gets trailing padding wrong, see #7798), but at least that interface seems to round-trip intervening padding bytes correctly:
```
>>> a = np.ones(4, np.dtype('i4,f8', align=True))
>>> a.dtype.descr
[('f0', '<i4'), ('', '|V4'), ('f1', '<f8')]
>>> memoryview(a).format
'T{i:f0:xxxxd:f1:}'
>>> np.array(memoryview(a)).dtype
dtype({'names':['f0','f1'], 'formats':['<i4','<f8'], 'offsets':[0,8], 'itemsize':16})
```

eric-wieser commented:
Can we not just store the `{'names':['f0','f1'], 'formats':['<i4','<f8'], 'offsets':[0,8], 'itemsize':16}` string in the file somehow? That seems pretty unambiguous to me.

ahaldane commented:
@eric-wieser, yeah, that's an option to investigate. I worry slightly that maybe the  `repr` of dtype objects might lose information, since maybe the repr wasn't coded with serialization in mind, while at least the pep3118 syntax was. We've previously discussed other cases in numpy where reprs in numpy aren't totally accurate (can't be used in `exec` to copy the array).

stefanv commented:
@ahaldane Mind if I rebase this to make CircleCI pass?

charris commented:
@ahaldane @eric-wieser Should we push this off to 1.16?

charris commented:
May need to rebase this to fix the circleci failure. Will now try the close/open route, but IIRC, that doesn't work.

charris commented:
This still seems controversial, what is the current status? If this is kicked off to 1.16 it seems that a forward port of  #10537 will be needed. There are also a number of PRs that reference this, my sense it that it may be best to kick this off one more release and concentrate on getting the dragon4 stuff in for 1.15.

charris commented:
@ahaldane @eric-wieser  ping.

ahaldane commented:
This one needs to be pushed off to 1.16 because the loadtxt/save fixes for padding bytes need to happen first.

I'll take a look at what is needed to forward port from 1.14.1.

charris commented:
> I'll take a look at what is needed to forward port from 1.14.1.

Great, thanks.

charris commented:
@ahaldane Just a ping, there are two issues blocking 1.15 that depend on resolving this.

ahaldane committed:
MAINT: push back multifield copy->view changes to 1.16

charris commented:
I changed the PR title, might want to check that. Does this also fix #10409?

charris commented:
Couple of failing tests.

ahaldane commented on numpy/core/src/multiarray/convert.c:
```diff
@@ -613,11 +613,15 @@ PyArray_View(PyArrayObject *self, PyArray_Descr *type, PyTypeObject *pytype)
+    flags = PyArray_FLAGS(self);
+    dtype = PyArray_DESCR(self);
+
+    if (type != NULL && !PyArray_EquivTypes(dtype, type) &&
```
The addition of `PyArrayEquivTypes` here, (the only real change), is so that code in `records.py` like `arr.view((np.record, arr.dtype))` doesn't complain. 

That kind of view doesn't change the memory layout so is save to keep in 1.16.

eric-wieser commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -1507,7 +1547,7 @@ _get_field_view(PyArrayObject *arr, PyObject *ind, PyArrayObject **view)
                 Py_DECREF(title);
             }
-            // disallow duplicate field indices
+            /* disallow duplicate field indices */
```
Why are we using C-style comments again? If we have a good reason, it seems our tests ought to be catching this...

ahaldane replied:
Yes, no test catches this. But it is in the style guide.

We did release 1.13 with these comments present and no-one complained, so maybe it is harmless. We fixed them in the 1.14 branch though.

Also, a grep shows that there is only one other place in numpy with a C++ comment, in `common.c`, which was also added by me in 2016. If we want to remove all C++ comments, I could fix that one into this PR too.

charris replied:
We do it because there used to be compilers that didn't accept the C++ style comments. We can probably drop this when we go to c99, although mixed comment styles can be a bit annoying. We should probably test for that, `-ansi` might detect that, although it could give us other errors. IIRC, it didn't work (many years ago).

ahaldane replied:
Right, although given that no-one complained in 1.13 it seems no-one building numpy uses incompatible compilers any more.

In any case, I just made the change in `common.c` in this PR, so now numpy is totally free of C++ comments.

eric-wieser commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -4796,9 +4796,25 @@ def test_field_names(self):
+            # assert_equal(b[['f1', 'f3']][0].tolist(), (2, (1,)))
+            with suppress_warnings() as sup:
+                sup.filter(FutureWarning,
+                           "Numpy has detected that you .*")
```
"selecting multiple fields" is a better pattern to look for here (possible with leading/trailing `.*`)

charris commented:
@ahaldane Should this be applied before or after the branch? The latter will leave master as is.

ahaldane commented:
Let's do it before the branch. It is easy to remove the unneeded parts when the time comes, and it will make the future update/change more explicit in the commit history.

eric-wieser commented on numpy/core/src/private/npy_import.h:
```diff
@@ -29,4 +29,18 @@ npy_cache_import(const char *module, const char *attr, PyObject **cache)
 }
 
+NPY_INLINE static PyObject *
+npy_import(const char *module, const char *attr)
```
Why was this function added?

ahaldane replied:
Oh... I thought it was used in `_multifield_view_to_copy`, but I must have removed that. Yes, I think it was related to the config variabel that I removed.

I'll make a followup to remove this.
